x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/FieldExtractorTestCase.java

```diff
@@ -0,0 +1,424 @@
+                    }
+                }
+            }
+            index.startObject("properties"); {
+                for (String type : types) {
+                    String fieldName = type + "_field";
+                    index.startObject(fieldName); {
+                        index.field("type", type);
+                        if (fieldProps != null) {
+                            Map<String, Object> fieldProp = fieldProps.get(fieldName);
+                            for (Entry<String, Object> prop : fieldProp.entrySet()) {
+                                index.field(prop.getKey(), prop.getValue());
+                            }
+                        }
+                    }
+                    index.endObject();
+                    
+                    if (withAlias) {
+                        // create two aliases - one within a hierarchy, the other just a simple field w/o hierarchy
+                        index.startObject(fieldName + "_alias"); {
+                            index.field("type", "alias");
+                            index.field("path", fieldName);
+                        }
+                        index.endObject();
+                        index.startObject("a.b.c." + fieldName + "_alias"); {
+                            index.field("type", "alias");
+                            index.field("path", fieldName);
+                        }
+                        index.endObject();
+                    }
+                }
+            }
+            index.endObject();
+        }
+        index.endObject();
+        index.endObject();
+
+        request.setJsonEntity(Strings.toString(index));
+        client().performRequest(request);
+    }
+    
+    private void index(String... docs) throws IOException {
+        Request request = new Request("POST", "/test/_bulk");
+        request.addParameter("refresh", "true");
+        StringBuilder bulk = new StringBuilder();
+        for (String doc : docs) {
+            bulk.append("{\"index\":{}\n");
+            bulk.append(doc + "\n");
+        }
+        request.setJsonEntity(bulk.toString());
+        client().performRequest(request);
+    }
+    
+    private Request buildRequest(String query) {
+        Request request = new Request("POST", RestSqlTestCase.SQL_QUERY_REST_ENDPOINT);
+        request.addParameter("error_trace", "true");
+        request.addParameter("pretty", "true");
+        request.setEntity(new StringEntity("{\"query\":\"" + query + "\",\"mode\":\"plain\"}", ContentType.APPLICATION_JSON));
+        
+        return request;
+    }
+    
+    private Map<String, Object> runSql(String query) throws IOException {
+        Response response = client().performRequest(buildRequest(query));
+        try (InputStream content = response.getEntity().getContent()) {
+            return XContentHelper.convertToMap(JsonXContent.jsonXContent, content, false);
+        }
+    }
+    
+    private JDBCType jdbcTypeFor(String esType) {
```

Comment:
Unfortunately, no.  `DataType` and `DataTypes` (belonging to `sql` project) are not visible in the `qa` project because `sql` is not a dependency. And `TypeUtils` is package protected.